src/datasources/accounts/counterfactual-safes/counterfactual-safes.datasource.ts

```diff
@@ -0,0 +1,145 @@
+import { IConfigurationService } from '@/config/configuration.service.interface';
+import { CacheRouter } from '@/datasources/cache/cache.router';
+import {
+  CacheService,
+  ICacheService,
+} from '@/datasources/cache/cache.service.interface';
+import { getFromCacheOrExecuteAndCache } from '@/datasources/db/utils';
+import { CounterfactualSafe } from '@/domain/accounts/counterfactual-safes/entities/counterfactual-safe.entity';
+import { CreateCounterfactualSafeDto } from '@/domain/accounts/counterfactual-safes/entities/create-counterfactual-safe.dto.entity';
+import { Account } from '@/domain/accounts/entities/account.entity';
+import { ICounterfactualSafesDatasource } from '@/domain/interfaces/counterfactual-safes.datasource.interface';
+import { ILoggingService, LoggingService } from '@/logging/logging.interface';
+import { Inject, Injectable, NotFoundException } from '@nestjs/common';
+import postgres from 'postgres';
+
+@Injectable()
+export class CounterfactualSafesDatasource
+  implements ICounterfactualSafesDatasource
+{
+  private readonly defaultExpirationTimeInSeconds: number;
+
+  constructor(
+    @Inject(CacheService) private readonly cacheService: ICacheService,
+    @Inject('DB_INSTANCE') private readonly sql: postgres.Sql,
+    @Inject(LoggingService) private readonly loggingService: ILoggingService,
+    @Inject(IConfigurationService)
+    private readonly configurationService: IConfigurationService,
+  ) {
+    this.defaultExpirationTimeInSeconds =
+      this.configurationService.getOrThrow<number>(
+        'expirationTimeInSeconds.default',
+      );
+  }
+
+  // TODO: the repository calling this function should:
+  // - check the AccountDataSettings to see if counterfactual-safes is enabled.
+  // - check the AccountDataType to see if it's active.
+  async createCounterfactualSafe(
+    account: Account,
+    createCounterfactualSafeDto: CreateCounterfactualSafeDto,
+  ): Promise<CounterfactualSafe> {
+    const [counterfactualSafe] = await this.sql<CounterfactualSafe[]>`
+      INSERT INTO counterfactual_safes 
+      ${this.sql([this.mapCreationDtoToRow(account, createCounterfactualSafeDto)])}
+      RETURNING *`;
+    const { key } = CacheRouter.getCounterfactualSafesCacheDir(account.address);
+    await this.cacheService.deleteByKey(key);
+    return counterfactualSafe;
+  }
+
+  async getCounterfactualSafe(id: string): Promise<CounterfactualSafe> {
```

Comment:
I would expect us to retrieve it by the predicted address, otherwise we are exposing the internal ID.

Reply:
We could also generate a unique uuid per CF Safe. 

And another thought: we need to keep in mind the endpoint to get this data would be authenticated.

Reply:
I would be more in favour of fetching by chain ID and address. It's easier for the clients imo.

Reply:
Done in [30f64bf](https://github.com/safe-global/safe-client-gateway/pull/1773/commits/30f64bf2cb6dd431bc2e863c75232d9fa5bb1467)

Reply:
I've added [a second PR](https://github.com/safe-global/safe-client-gateway/pull/1776) to type the datasource arguments, as it's otherwise easy to make an error while passing parameters between the layers. This aligns the implementation more with other classes in the codebase.